text/0000-constrained-naked.md

```diff
@@ -0,0 +1,142 @@
+
+# Motivation
+
+Naked functions have long been a feature of compilers. These functions are typically defined as normal functions in every regard, except that the compiler does not emit the function prologue and epilogue. Rust's early attempt to support this feature ([RFC 1201](https://github.com/rust-lang/rfcs/blob/master/text/1201-naked-fns.md)) mostly copied the existing compiler behaviors.
+
+However, naked functions are often avoided in practice because their behavior is not well defined. The root cause of this problem is that naked functions are defined by negation: they are functions which lack a prologue and epilogue. Unfortunately, functions that lack a prologue and epilogue present a number of complicated problems that the compiler needs to solve and developers need to work around. And there is a long history of compilers and developers getting this wrong.
+
+This document seeks to define naked functions in a much more constrained, positivistic way. In doing so, naked functions can become more useful.
+
+# Naked function definition
+
+A naked function is a type of FFI function with a defined calling convention and a body which contains only assembly code which can rely upon the defined calling convention.
+
+A naked function is identified by the `#[naked]` attribute and:
+1. must specify a calling convention besides `extern "Rust"`.
+1. must define only FFI-safe arguments and return types.
+1. must not specify the `#[inline]` or `#[inline(always)]` attribute.
+1. must have a body which contains only a single `asm!()` statement which:
+    1. may be wrapped in an `unsafe` block.
+    1. must not contain any operands except `const` or `sym`.
+    1. must contain the `noreturn` option.
+    1. must not contain any other options except `att_syntax`.
+    1. must ensure that the requirements of the calling convention are followed.
+
+In exchange for the above constraints, the compiler commits to:
+1. produce a clear error if any of the above rules are violated.
+1. never inline the function (implicit `#[inline(never)]`).
+1. emit no instructions to the function body that are not contained in the `asm!()` statement.
+
+As a (weaker) correlary to the last compiler commitment, since the compiler generates no additional instructions, the initial state of all registers in the `asm!()` statement conform to the specified calling convention.
+
+# Custom Calling Convention
+
+In order to expand the usefulness of naked functions, this document also define a new calling convention, `extern "custom"`, that can be used only with naked functions. Functions with `extern "custom"` are not callable from Rust and must be marked as`unsafe`. The function definition must contain no arguments or return type. For example:
```

Comment:
To me,

> that can be used only with naked functions

means that you can't declare an `extern "custom"` function that is defined in a different language.  I assume you intended for this to be allowed, and maybe the wording should change.

As for the `unsafe` issue, it occurs to me that `extern "custom"` cheats the type system a bit.  It's possible and even likely for multiple `extern "custom"` functions that appear identically typed to Rust to have different custom calling conventions.  This could be awkwardly addressed by something like `extern "custom/x86_idt"` and `extern "custom/x86_syscall"`, or it could be addressed by encouraging people who pass around pointers to `extern "custom"` functions to wrap them in more descriptive structs.  To that end, perhaps `unsafe` is the right choice.

As another way of looking at this, I can write a safe function (safe to call and implemented without `unsafe`) that takes a pointer to a function with a known calling convention and calls it.  The called function may itself be problematic, but one might argue that this is a problem with the callee.  But I can't write a function that is safe to call and takes an `extern "custom"` pointer because nothing verifies that the passed pointer is the correct type of custom.

(As an aside, it's not quite clear to me what "memory safety" means in the context of an operating system.  Dereferencing a wild pointer is obviously memory unsafe, but what about mis-programming the CPU or associated hardware in a way that causes crashes?  I suppose this boils down to the preference of whomever designs the OS.)

Reply:
Yes, I think we should definitely think about the calling convention as a "function type." This is, I think, behind my intuition that these should be marked as `unsafe`. Basically, `extern "custom"` is a form of type erasure.

I like the idea of a `custom` prefix which allows you to declare your own name. This sort of adds type safety. Unless there's a collision, of course.

Reply:
> I like the idea of a `custom` prefix which allows you to declare your own name. This sort of adds type safety. Unless there's a collision, of course.

What about going all out and having non-string ABIs -- similar to generics in that the ABI is a type parameter to the generic class of functions. `extern "MyABI"` would be syntax sugar for `extern<core::abi::MyABI>` where MyABI is translated by replacing all `-` with `_`. So `extern "system-unwind"` would be syntax sugar for `extern<core::abi::system_unwind>`.

```rust
// declare a custom ABI
#[derive(ExternABI)]
struct MyABI;

// use a custom ABI
fn f(fn_ptr: unsafe extern<MyABI> fn()) {}

fn generic_over_abi<ABI: ExternABI>(fn_ptr: unsafe extern<ABI> fn()) {}

// both the same type
type ExternCFnPtr = unsafe extern<core::abi::C> fn();
type ExternCFnPtr2 = unsafe extern "C" fn();
```

Reply:
I feel that this is over-extending the scope of the RFC. Simple newtype wrappers around an `extern "custom"` function pointer are enough to build type-safe abstractions, there is no need to add complexity to the language for such a niche feature.

Regarding `unsafe`, it really doesn't make a difference in practice since the only difference between a `fn` and `unsafe fn` is that the latter can only be called from within an `unsafe` block. This doesn't apply to `extern "custom"` since it can't be called from Rust at all.

Reply:
> I feel that this is over-extending the scope of the RFC. Simple newtype wrappers around an `extern "custom"` function pointer are enough to build type-safe abstractions, there is no need to add complexity to the language for such a niche feature.

Ok, though my proposal is also useful for writing code that is generic over the ABI, which is useful when trying to handle generic `fn` types. This is more related to the RFC for adding a trait that is implemented by all `fn` types, which I can't find at the moment.

Reply:
@Amanieu To me, there is some difference in meaning between having the function `unsafe` or not. If I see:
- `extern "C" { fn foo(f: extern "custom" fn()); }`, then I expect any non-unsafe `extern "custom"` function to be possible to put there without memory safety impact — meaning that non-unsafe `extern "custom"` functions are necessarily functions that never return and that don't assume anything about the caller, to make it possible to have such a signature
- `extern "C" { fn foo(f: unsafe extern "custom" fn()); }`, then I expect `foo` to run `f` in some kind of virtual machine to protect itself against misuse, as it cannot know for sure what ABI to call `f` in
- `extern "C" { unsafe fn foo(f: [unsafe] extern "custom" fn()); }`, then I read `foo`'s documentation on what kinds of `extern "custom"` functions it accepts

So I do believe that there is a semantic difference between `unsafe` and non-`unsafe` `extern "custom"` functions, but the non-unsafe `extern "custom"` functions are restricted to the set of functions that never return and don't assume anything from their caller (apart from what all rust programs assume by default, eg. “there is a stack and it doesn't overlap with the heap”)

I'd be interested in other views of which `extern "custom"` functions should or should not be `unsafe`, that said